### ISSUE TITLE
Organize urls better

### ISSUE THREAD

**Issue by jmoldow** [learning-unlimited/ESP-Website#1292]
Our system of dealing with urls is not as great as it should be. I would like to see us eventually get to a more organized url system, by:
- Using the `url()` function (https://docs.djangoproject.com/en/1.4/topics/http/urls/#django.conf.urls.url), in order to
- Use named urls (https://docs.djangoproject.com/en/1.4/topics/http/urls/#naming-url-patterns)
- Split central monolithic URLconf into app-specific URLconfs, which are included in the root URLconf `esp/esp/urls.py` (https://docs.djangoproject.com/en/1.4/topics/http/urls/#including-other-urlconfs)
- Stop hard-coding urls, and instead use named urls to call urlresolver utility functions (https://docs.djangoproject.com/en/1.4/topics/http/urls/#django-core-urlresolvers-utility-functions), especially `reverse()` and the `{% url %}` template tag.
- Define `get_absolute_url()` (https://docs.djangoproject.com/en/1.4/ref/models/instances/#get-absolute-url), using the `@permalink` decorator (https://docs.djangoproject.com/en/1.4/ref/models/instances/#django.db.models.permalink), on models with canonical urls. This won't apply to most models, but it applies to `ESPUser`, some of the various resource models, `customforms.Form`, `TeacherBio`, and QSD, and maybe others.
- Even in places where it doesn't make sense to define `get_absolute_url()`, it might still be possible to write helper methods that use the `@permalink` decorator to make it easy to get canonical urls without hardcoding them. I can see this being useful on `Program` and `ProgramModule`s.

With regards to named urls: with the current architecture, we won't be able to take 100% advantage of named urls. This is because a large number of our view functions are hidden behind the program module framework. As a nearly-good-enough workaround, when reversing urls for program module views, reverse `esp.web.views.main.program`, with the `module` argument set to the view name. However, the true fix would be to refactor the program module mechanism to somehow use URLconfs, which can then be defined with named urls.

**Comment by imran-techvoyager** [comment]
Hi all @jmoldow @mgersh @willgearty I did a current-state audit of URLs across the codebase to see what work remains under #1292.

PR #2121 addressed the deprecated patterns() and string-view arguments, and partially split URLconfs. That part looks complete.

What still remains is primarily:

Many hardcoded URLs in Python views (e.g. HttpResponseRedirect with string formatting).

Very limited use of named URLs and reverse().

get_absolute_url() methods that build strings instead of using reverse().

Before touching the deeper architectural question around program module dispatch (which likely requires design discussion), I’d like to propose starting with a narrow, low-risk cleanup:

Phase 1 (mechanical, no behavior change):

Add name= to currently unnamed patterns in app-specific URLconfs.

Replace hardcoded redirects in Python (about ~29 occurrences) with reverse().

Update get_absolute_url() implementations to use reverse() instead of string formatting.

This would:

Reduce brittle string formatting

Improve safety of future URL refactors

Move incrementally toward the original vision of #1292

Avoid touching the program module catch-all routing (for now)

If that direction sounds reasonable, I can open a focused PR scoped only to Phase 1.

**Comment by willgearty** [comment]
Yes, that Phase 1 sound like a good first step towards this PR. Thanks @imran-techvoyager

**Comment by imran-techvoyager** [comment]
@willgearty , My Phase 1 work (#4465) is merged. The remaining scope is the template-side refactoring tracked in #5261, which @Janvi-kapoor is working on. Unassigning myself, so that i can take up other issues

**Comment by Janvi-kapoor** [comment]
Hi @willgearty, @imran-techvoyager, thank you for the reviews!

Since Batch 1 (#5322) and Batch 2 (#5443) are now fully standardized with named URL patterns and the feedback has been addressed, they are ready for merge.

I currently have my university exams going on, but I’ll start working on Batch 3 as soon as I get a moment. Looking forward to getting these merged so I can continue the refactor!
